src/main/java/ch/njol/skript/lang/SkriptParser.java

```diff
@@ -263,7 +264,46 @@ private <T extends SyntaxElement> T parse(Iterator<? extends SyntaxElementInfo<?
 								x = x2;
 							}
 							T t = info.c.newInstance();
-							if (t.init(res.exprs, i, getParser().getHasDelayBefore(), res)) {
+							ParserInstance parser = getParser();
+							boolean error = false;
+							Class<? extends Structure>[] usableStructures = t.getUsableStructures();
+							if (parser.getCurrentStructure() != null) {
+								if (usableStructures.length > 0) {
+									error = true;
+									for (Class<? extends Structure> clazz : usableStructures) {
+										if (parser.isCurrentStructure(clazz)) {
+											error = false;
+											break;
+										}
+									}
+									if (error) {
+										Skript.error("'" + res.expr + "' cannot be used in " + Utils.a(parser.getCurrentStructure().toString(null, false)) + " structure.");
+									}
+								}
+							} else if (usableStructures.length > 0) {
+								error = true;
+								Skript.error("'" + res.expr + "' cannot be used outside of the allowed structures.");
+							}
+							Class<? extends Section>[] usableSections = t.getUsableSections();
+							List<TriggerSection> sections = parser.getCurrentSections();
+							TriggerSection current = sections.isEmpty() ? null : sections.get(sections.size() - 1);
+							if (!error && usableSections.length > 0) {
+								error = true;
+								if (current == null) {
```

Comment:
since we simplified the currentSection check this could likely become `parser.getCurrentSections().isEmpty()`